File: Raster.Samples/native-example/scripts/modules/opengl.js

var _cpp = cpp;

/** Exports for CommonJs modules  */
exports.GL_COLOR_BUFFER_BIT = 16384;
exports.GL_DEPTH_BUFFER_BIT = 256;
exports.GL_PROJECTION = 5889;
exports.GL_DEPTH_TEST = 2929;
exports.GL_MODELVIEW = 5888;
exports.GL_TRIANGLES = 4;
exports.GL_QUADS = 7;

exports.glBegin = function(type) {
    _cpp.glBegin(type);
};

exports.glEnd = function() {
    _cpp.glEnd();
}

exports.glClearColor = function(r, g, b, a) {
    _cpp.glClearColor(r, g, b, a);
}

exports.glClear = function(buffers) {
    _cpp.glClear(buffers);
}

exports.glVertex2f = function(x, y) {
    _cpp.glVertex2f(x, y);
}

exports.gluPerspective = function(fov, aspect, near, far) {
    _cpp.gluPerspective(fov, aspect, near, far);
}

exports.glVertex3f = function (x, y, z) {
    _cpp.glVertex3f(x, y, z);
}

exports.glColor3f = function (r, g, b) {
    _cpp.glColor3f(r, g, b);
}

exports.glTranslatef = function(x, y, z) {
    _cpp.glTranslatef(x, y, z);
};

exports.glRotatef = function(rotate, x, y, z) {
    _cpp.glRotatef(rotate, x, y, z);
}

exports.glMatrixMode = function(matrixMode) {
    _cpp.glMatrixMode(matrixMode);
}

exports.glLoadIdentity = function() {
    _cpp.glLoadIdentity();
}

exports.glEnable = function(enable) {
    _cpp.glEnable(enable);
}

exports.default = {};
for (var key in exports) {
    if (key !== 'default') {
        exports.default[key] = exports[key];
    }
}

File: Raster.Core/OpenGL.cpp

#include "OpenGL.h"
#include "gl/GLU.h"
#include "gl/gl.h"
#include "JsExtensions.h"

void raster::glClearColor(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 4)
    {
        auto r = args[0].As<v8::Number>()->Value();
        auto g = args[1].As<v8::Number>()->Value();
        auto b = args[2].As<v8::Number>()->Value();
        auto a = args[3].As<v8::Number>()->Value();

        ::glClearColor(r, g, b, a);
    }
}

void raster::glClear(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 1)
    {
        ::glClear(args[0]->NumberValue());
    }
}

void raster::glBegin(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 1)
    {
        ::glBegin(static_cast<GLenum>(args[0]->NumberValue()));
    }
}

void raster::glColor3(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 3)
    {
        auto r = args[0].As<v8::Number>()->Value();
        auto g = args[1].As<v8::Number>()->Value();
        auto b = args[2].As<v8::Number>()->Value();

        glColor3f(r, g, b);
    }
}

void raster::glVertex2(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 2)
    {
        auto x = args[0].As<v8::Number>()->Value();
        auto y = args[1].As<v8::Number>()->Value();

        glVertex2f(x, y);
    }
}

void raster::glVertex3(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 3)
    {
        auto x = args[0].As<v8::Number>()->Value();
        auto y = args[1].As<v8::Number>()->Value();
        auto z = args[2].As<v8::Number>()->Value();

        glVertex3f(x, y, z);
    }
}

void raster::glLoadIdentity(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    ::glLoadIdentity();
}

void raster::glMatrixMode(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 1)
    {
        ::glMatrixMode(args[0].As<v8::Number>()->Value());
    }
}

void raster::glTranslate(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 3)
    {
        auto x = args[0].As<v8::Number>()->Value();
        auto y = args[1].As<v8::Number>()->Value();
        auto z = args[2].As<v8::Number>()->Value();

        glTranslatef(x, y, z);
    }
}

void raster::glRotate(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 4)
    {
        auto r = args[0].As<v8::Number>()->Value();
        auto x = args[1].As<v8::Number>()->Value();
        auto y = args[2].As<v8::Number>()->Value();
        auto z = args[3].As<v8::Number>()->Value();

        glRotatef(r, x, y, z);
    }
}

void raster::glPerspective(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 4)
    {
        auto fov = args[0].As<v8::Number>()->Value();
        auto aspect = args[1].As<v8::Number>()->Value();
        auto min = args[2].As<v8::Number>()->Value();
        auto max = args[3].As<v8::Number>()->Value();

        gluPerspective(fov, aspect, min, max);
    }
}

void raster::glEnd(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    ::glEnd();
}

void raster::glEnable(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    if (args.Length() == 1)
    {
        ::glEnable(args[0].As<v8::Integer>()->Value());
    }

}

void raster::registerOpenGL(v8::Local<v8::Object>& object) 
{
    AttachFunction(object, "gluPerspective", glPerspective);
    AttachFunction(object, "glEnable", glEnable);
    AttachFunction(object, "glMatrixMode", glMatrixMode);
    AttachFunction(object, "glLoadIdentity", glLoadIdentity);
    AttachFunction(object, "glTranslatef", glTranslate);
    AttachFunction(object, "glRotatef", glRotate);
    AttachFunction(object, "glClearColor", glClearColor);
    AttachFunction(object, "glVertex3f", glVertex3);
    AttachFunction(object, "glVertex2f", glVertex2);
    AttachFunction(object, "glColor3f", glColor3);
    AttachFunction(object, "glClear", glClear);
    AttachFunction(object, "glBegin", glBegin);
    AttachFunction(object, "glEnd", glEnd);
}